src/posix/platform/resolver.cpp

```diff
@@ -0,0 +1,293 @@
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "resolver.hpp"
+
+#include "platform-posix.h"
+
+#include <openthread/logging.h>
+#include <openthread/message.h>
+#include <openthread/udp.h>
+#include <openthread/platform/dns.h>
+
+#include "common/code_utils.hpp"
+
+#include <arpa/inet.h>
+#include <arpa/nameser.h>
+#include <cassert>
+#include <netinet/in.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+
+#include <fstream>
+#include <string>
+
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+
+namespace {
+constexpr char kResolveConfLocation[] = "/etc/resolv.conf";
+constexpr char kNameserverItem[]      = "nameserver";
+} // namespace
+
+extern ot::Posix::Resolver gResolver;
+
+namespace ot {
+namespace Posix {
+
+void Resolver::Init(void)
+{
+    memset(mUpstreamTransaction, 0, sizeof(mUpstreamTransaction));
+
+    LoadDnsServerListFromConf();
+}
+
+void Resolver::LoadDnsServerListFromConf(void)
+{
+    std::string   line;
+    std::ifstream fp;
+
+    fp.open(kResolveConfLocation);
+    if (fp.bad())
+    {
+        otLogCritPlat("Cannot read %s for domain name servers, default to 127.0.0.1", kResolveConfLocation);
+        mUpstreamDnsServerCount = 1;
+        assert(inet_pton(AF_INET, "127.0.0.1", &mUpstreamDnsServerList[0]) == 1);
+        ExitNow();
+    }
+
+    while (std::getline(fp, line) && mUpstreamDnsServerCount < kMaxUpstreamServerCount)
+    {
+        if (line.find(kNameserverItem, 0) == 0)
+        {
+            in_addr_t addr;
+
+            if (inet_pton(AF_INET, &line.c_str()[sizeof(kNameserverItem)], &addr) == 1)
+            {
+                otLogCritPlat("Got nameserver #%d: %s", mUpstreamDnsServerCount,
+                              &line.c_str()[sizeof(kNameserverItem)]);
+                mUpstreamDnsServerList[mUpstreamDnsServerCount] = addr;
+                mUpstreamDnsServerCount++;
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void Resolver::Query(otPlatDnsUpstreamQuery *aTxn, const otMessage *aQuery)
+{
+    char        packet[kMaxDnsMessageSize];
+    otError     error  = OT_ERROR_NONE;
+    uint16_t    length = otMessageGetLength(aQuery);
+    sockaddr_in serverAddr;
+
+    Transaction *txn;
+
+    VerifyOrExit(length <= kMaxDnsMessageSize, error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(otMessageRead(aQuery, 0, &packet, sizeof(packet)) == length, error = OT_ERROR_NO_BUFS);
+
+    txn = AllocateTransaction(aTxn);
+    VerifyOrExit(txn != nullptr, error = OT_ERROR_NO_BUFS);
+
+    serverAddr.sin_family = AF_INET;
+    serverAddr.sin_port   = htons(53);
+    for (int i = 0; i < mUpstreamDnsServerCount; i++)
```

Comment:
We send the query to multiple servers?
Do we accept the first one which responsds ?

Reply:
Yes, we sends the query to multiple servers, and accepts the first one and drop the others.
This behavior just like `--all-server` for `dnsmasq`